File: src/Token.hs
module Token where

data TokenType =  PLUS      | MINUS  | STAR   | SLASH | RSLASH | EXPCLAMATION | POWER  
                | BIND      | EQUAL  | DEQUAL | LESS  | GREAT  | LEQUAL       | GEQUAL 
                | FALSE     | TRUE   | IF     | THEN  | ELSE   | AND          | OR     | NOT 
                | NAME      | LAMBDA 
                | INT       | FLOAT  
                | LPAREN    | RPAREN 
                | SEMICOLON | EOF    
                | NIL                  deriving Show

data Category = None | Num | Name | InfixOp

data Token    = Token { token :: TokenType, content :: String, suptype :: Category }

instance Show Token where
    show (Token token content _) = 
        " {token = " ++ (show token) ++ 
        ", content = " ++ (show content) ++ "} "

File: src/main/scala/Token.scala
sealed abstract class TokenType

class Token(val token: TokenType, val content: String) {
  override def toString: String = "{ " + token + ", " + content + " }"
}

object Token {
  case object NAME extends TokenType
  case object ATOM extends TokenType
  case object LAMBDA extends TokenType
  case object DELIMITER extends TokenType
  case object QUOTE extends TokenType
  case object RETURN extends TokenType
  case object NUM extends TokenType
  case object CHAR extends TokenType
  case object STRING extends TokenType
  case object LPAREN extends TokenType
  case object RPAREN extends TokenType
  case object SEMICOLON extends TokenType
  case object EOF extends TokenType
}